lasagne/layers/normalization.py

```diff
@@ -101,3 +115,233 @@ def get_output_for(self, input, **kwargs):
+
+        if axes == 'auto':
+            # default: normalize over all but the second axis
+            axes = (0,) + tuple(range(2, len(self.input_shape)))
+        elif isinstance(axes, int):
+            axes = (axes,)
+        self.axes = axes
+
+        self.epsilon = epsilon
+        self.alpha = alpha
+        if nonlinearity is None:
+            nonlinearity = nonlinearities.identity
+        self.nonlinearity = nonlinearity
+        self.mode = mode
+
+        # create parameters, ignoring all dimensions in axes
+        shape = [size for axis, size in enumerate(self.input_shape)
+                 if axis not in self.axes]
+        if any(size is None for size in shape):
+            raise ValueError("BatchNormLayer needs specified input sizes for "
+                             "all axes not normalized over.")
+        self.beta = self.add_param(beta, shape, 'beta',
+                                   trainable=True, regularizable=False)
+        self.gamma = self.add_param(gamma, shape, 'gamma',
+                                    trainable=True, regularizable=True)
+        self.mean = self.add_param(mean, shape, 'mean',
+                                   trainable=False, regularizable=False)
+        self.var = self.add_param(var, shape, 'var',
+                                  trainable=False, regularizable=False)
+
+    def get_output_for(self, input, deterministic=False, **kwargs):
+        input_mean = input.mean(self.axes)
+        input_var = input.var(self.axes)
+
+        # Decide whether to use the stored averages or mini-batch statistics
+        use_averages = kwargs.get('batch_norm_use_averages',
+                                  deterministic)
+        if use_averages:
+            mean = self.mean
+            var = self.var
+        else:
+            mean = input_mean
+            var = input_var
+
+        # Decide whether to update the stored averages
+        update_averages = kwargs.get('batch_norm_update_averages',
+                                     not deterministic)
+        if update_averages:
+            # Trick: To update the stored statistics, we create memory-aliased
+            # clones of the stored statistics:
+            running_mean = theano.clone(self.mean, share_inputs=False)
+            running_var = theano.clone(self.var, share_inputs=False)
+            # set a default update for them:
+            running_mean.default_update = ((1 - self.alpha) * running_mean +
+                                           self.alpha * input_mean)
+            running_var.default_update = ((1 - self.alpha) * running_var +
+                                          self.alpha * input_var)
+            # and make sure they end up in the graph without participating in
+            # the computation (this way their default_update will be collected
+            # and applied, but the computation will be optimized away):
+            mean += 0 * running_mean
```

Comment:
This is... interesting :) But it's pretty cool that this works. Have you verified in any way that the computation is indeed optimized away? It's not always clear which optimizations are applied and which aren't...